connectors/flink/src/main/scala/io/delta/flink/internal/KernelDeltaLogDelegator.scala

```diff
@@ -0,0 +1,107 @@
+      tableClient,
+      -1,
+      new SnapshotManager(),
+      java.util.Optional.empty())
+
+/**
+ * We want to be able to construct an OptimisticTransactionImpl that uses a delta log and a snapshot
+ * provided by the Delta Kernel. OptimisticTransactionImpl takes a DeltaLogImpl and SnapshotImpl
+ * internally, so we need classes that extend those, and this is the one for DeltaLogImpl. It
+ * provides features used by flink+startTransaction.
+ */
+class KernelDeltaLogDelegator(
+    tableClient: DefaultTableClient,
+    table: TableImpl,
+    hadoopConf: Configuration,
+    logPath: Path,
+    dataPath: Path,
+    clock: Clock)
+  extends DeltaLogImpl(hadoopConf, logPath, dataPath, clock) {
+
+  val standaloneDeltaLog = new DeltaLogImpl(hadoopConf, logPath, dataPath, clock)
+  var currKernelSnapshot: Option[KernelSnapshotDelegator] = None
+
+  override def snapshot(): StandaloneSnapshotImpl = { // but is actually a KernelSnapshotDelegator
+    if (currKernelSnapshot.isEmpty) { update() }
+    return currKernelSnapshot.get
+  }
+
+  override def update(): StandaloneSnapshotImpl = { // but is actually a KernelSnapshotDelegator
+    // get latest snapshot via kernel
+    val kernelSnapshot = try {
+      table.getLatestSnapshot(tableClient).asInstanceOf[SnapshotImplKernel]
+    } catch {
+      case e: TableNotFoundException =>
+        new InitialKernelSnapshotImpl(logPath, dataPath, tableClient)
+    }
+    currKernelSnapshot = Some(new KernelSnapshotDelegator(
+      kernelSnapshot,
+      hadoopConf,
+      logPath,
+      kernelSnapshot.getVersion(tableClient), // note: tableClient isn't used
+      this,
+      standaloneDeltaLog
+    ))
+    currKernelSnapshot.get
+  }
+
+  override def startTransaction(): io.delta.standalone.OptimisticTransaction = {
+    val snapshot = update()
+    new OptimisticTransactionImpl(this, snapshot)
+  }
+
+  override def tableExists: Boolean = snapshot.version >= 0
+
+  override def getChanges(startVersion: Long, failOnDataLoss: Boolean): java.util.Iterator[VersionLog] = {
+    logWarning("KernelDeltaLogDelegator falling back to DeltaLogImpl for getChanges")
+    standaloneDeltaLog.getChanges(startVersion, failOnDataLoss)
+  }
+
+  override def getSnapshotForVersionAsOf(version: Long): StandaloneSnapshotImpl = {
+    logWarning("KernelDeltaLogDelegator falling back to DeltaLogImpl for getSnapshotForVersionAsOf")
+    standaloneDeltaLog.getSnapshotForVersionAsOf(version)
+  }
+  override def getSnapshotForTimestampAsOf(timestamp: Long): StandaloneSnapshotImpl = {
+    throw new RuntimeException()
+  }
+  override def getCommitInfoAt(version: Long): CommitInfoJ = {
+    throw new RuntimeException()
+  }
+}
+
+object KernelDeltaLogDelegator {
+  def forTable(hadoopConf: Configuration, dataPath: String): KernelDeltaLogDelegator = {
+    val tableClient = DefaultTableClient.create(hadoopConf)
+    val table = Table.forPath(tableClient, dataPath).asInstanceOf[TableImpl]
```

Comment:
Perhaps:
- create the `standaloneDeltaLog` first using `io.delta.standalone.DeltaLog.forTable(...)`
- this will create the delta log folder, if it doesn't already exist!
- do this before the `Table.forPath` call
- then pass that in as a param into `KernelDeltaLogDelegator`

Reply:
yep, makes sense